# Extract kiwi genes

Now that we have genotype calls for all sites in the genome for all samples, we can extract our sequences of interest.

First, make a list.

```bash
cat > samples
```

Here is the list of samples with the name a it appears in the vcf and the shorter name, separated by a space:

```
/home/0_GENOMES5/kiwi/0_mapped/australis__Haast__KW36__RA0997_sorted.bam australis__Haast__KW36__RA0997
/home/0_GENOMES5/kiwi/0_mapped/australis__NorthFiordland__KW41__R32961_sorted.bam australis__NorthFiordland__KW41__R32961
/home/0_GENOMES5/kiwi/0_mapped/australis__SouthFiordland__KW50__RA0202_sorted.bam australis__SouthFiordland__KW50__RA0202
/home/0_GENOMES5/kiwi/0_mapped/australis__StewartIsland__KW51__RA0891_sorted.bam australis__StewartIsland__KW51__RA0891
/home/0_GENOMES5/kiwi/0_mapped/haastii__haastii__KW26__CD830_sorted.bam haastii__haastii__KW26__CD830
/home/0_GENOMES5/kiwi/0_mapped/mantelli__Coromandel__KW09__R32852_sorted.bam mantelli__Coromandel__KW09__R32852
/home/0_GENOMES5/kiwi/0_mapped/mantelli__Eastern__KW11__NIBKOpo1_sorted.bam mantelli__Eastern__KW11__NIBKOpo1
/home/0_GENOMES5/kiwi/0_mapped/mantelli__Northland__KW04__41_sorted.bam mantelli__Northland__KW04__41
/home/0_GENOMES5/kiwi/0_mapped/mantelli__Taranaki__KW18__kiwiCarcass1_sorted.bam mantelli__Taranaki__KW18__kiwiCarcass1
/home/0_GENOMES5/kiwi/0_mapped/owenii__Kapiti__KW24__O20581_sorted.bam owenii__Kapiti__KW24__O20581
/home/0_GENOMES5/kiwi/0_mapped/rowi__Okarito__KW32__R32934_sorted.bam rowi__Okarito__KW32__R32934
```

We can use ```bcftools consensus``` to obtain the CDS sites (coding sequences) using a VCF file. Importantly, the -M and --absent options should be set to avoid having missing data replaced with the reference allele (this requires bcftools version 1.11+).

```bash
mkdir genomes
#create a consensus genome sequence for each sample
cat samples | parallel --colsep " " 'bcftools consensus --fasta-ref /kiwi_ref_genome.fna --sample {1} -M N -a N -H 1 Kiwi_11genomes.allsites.vcf.gz > genomes/{2}.fa'

mkdir filtered_genomes
#create a consensus genome sequence for each sample
cat samples | parallel --colsep " " 'bcftools consensus --fasta-ref kiwi_ref_genome.fna --sample {1} -M N -a N -H 1 Kiwi_11genomes.min5max21.vcf.gz > filtered_genomes/{2}.fa'

#Finally, it turns out that the samples were mapped to the genbank assembly but the annotation file is for the refseq assembly! Luckily the assemblies are identical, except that the names of the scaffolds are different. We just need to rename everything
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/003/343/035/GCF_003343035.1_aptRow1/GCF_003343035.1_aptRow1_assembly_report.txt #lists the translation between Refseq and Genbank names
grep "RefSeq assembly and GenBank assemblies identical:" GCF_003343035.1_aptRow1_assembly_report.txt
# RefSeq assembly and GenBank assemblies identical: yes
grep -v "^#" GCF_003343035.1_aptRow1_assembly_report.txt | cut -f 5,7 > Genbank_Refseq_namingtable #needs to be a 2-column, tab-separated file, in which col1=oldname and col2=newname to be replaced
cp -r filtered_genomes temp_filtered_genomes #making a copy in case something goes wrong
conda activate maker
cat samples | parallel --colsep " " time maker/bin/map_fasta_ids Genbank_Refseq_namingtable filtered_genomes/{2}.fa
#note if you had already indexed the fastas, you must now index them again!

```
* "-s, --sample <name>        apply variants of the given sample"
* "-M, --missing <char>       output <char> instead of skipping the missing genotypes"
* "-a, --absent <char>        replace positions absent from VCF with <char>"

Now we have a "whole genome" sequence for each sample, and we just need to extract the sequences of interest.

## Obtain bed file for each gene
  
Firstly, we need a bed file listing the location of each gene in the genome. Critically, this must list the coordinates for the coding sequences (CDS), without the introns or untranslated regions, since we will need to be able to translate them straight into protein sequence.
 
```bash
export PATH=/home/0_PROGRAMS/bedops/bin:$PATH
#get BED files for each protein
#loop through the list of genes. For each gene, look in the gff file for lines for the gene of interest, where column three indicates that that line pertains to the CDS. Then convert it to bed format.
mkdir bed
cat Aptrow.longest_RNA.txt | while read gene ; do grep "$gene" kiwi_ref_genome/genomic.gff | awk '($3 == "CDS")' - | bedops/bin/convert2bed -i gff - > bed/"$gene".bed ; done

#flip the order of the exons for genes on the minus strand, so that they are in the correct order relative to the gene, not the chromosome.
#first, get a file listing whether each gene is on the plus or minus strand
cat Aptrow.longest_RNA.txt | while read gene ; do head -n 1 bed/"$gene".bed | cut -f 6 | xargs -I {} echo "$gene" {} >> strandedness ; done
#reverse the order of the exons in the bed file for genes on the minus strand
cat strandedness | while read gene strandedness ; do if [[ "$strandedness" == "-" ]]; then sort --numeric-sort --reverse -k 2 bed/"$gene".bed > temp && mv temp bed/"$gene".bed ; fi ; done

```

## Extract CDS from each sample

  Now we can use bedtools to extract the CDS from each sample using the bed files. 
  
```bash
mkdir genes
  
#bedtools will extract a sequence for each of the CDS exons listed in the file, producing separate fasta entries for each exon
#then, we can process that file right away: remove the headers, remove the newlines (concatenating the exons), add a newline to the end, then add a header. That gives a single fasta entry for the whole CDS
#then just concatenate that sequence to a single fasta that will hold the gene sequence for all samples. Make sure the unique sample name makes it into the fasta header so you can tell which is which!

#first, start it out with just the reference genome
#cat Aptrow.longest_RNA.txt | while read gene ; do bedtools getfasta -fi kiwi_ref_genome.fna -bed bed/"$gene".bed | sed 's/>.*$//g' | tr -d '\n' | sed 's/$/\n/g' | sed "s/^/>$sample\n/g" >> genes/"$gene".fa ; done

#loop through each sample and each gene
cat samples | cut -f 2 -d " " | while read sample ; do time cat Aptrow.longest_RNA.txt | while read gene ; do bedtools getfasta -s -fi "$sample".fa -bed bed/"$gene".bed | sed 's/>.*$//g' | tr -d '\n' | sed 's/$/\n/g' | sed "s/^/>$sample\n/g" >> genes/"$gene".fa ; done ; done

#cat samples | cut -f 2 -d " " | while read sample ; do time cat Aptrow.longest_RNA.txt | parallel bedtools getfasta -fi "$sample".fa -bed bed/{1}.bed '|' sed 's/>.*$//g' '|' tr -d '\n' '|' sed 's/$/\n/g' '|' sed "s/^/>$sample\n/g" '>>' genes/{1}.fa ; done

```
* `-fi`: fasta input sequence
* `-s`:	"Force strandedness. If the feature occupies the antisense strand, the sequence will be reverse complemented. Default: strand information is ignored."

Now we have fasta files containing (unaligned) gene sequences for each annotated kiwi gene! 
This should be aligned (will not need much alignment as these are extracted from artificial genomes based on the same reference); they should be trimmed to remove samples with lots of missing data. Do not trim the alignment in any way that could compromise the reading frame!

# Appendix: choice of gene isoforms
There are multiple isoforms available for some genes. Choose the longest isoform of each gene.

## Select sequences
I have already made a list of the longest isoforms for each gene annotated for Apteryx. I will use this list. It lists the protein names, so I just need to translate these into a list of RNA names that will be used for the CDS sequences.

```bash
cat Aptrow.longest_AA.txt | while read protein ; do grep "$protein" kiwi_ref_genome/genomic.gff | head -n 1 | sed 's/^.*;Parent=//g' | sed 's/;.*$//g' >> Aptrow.longest_RNA.txt ; done
```
